### DISCUSSION TITLE
Refactor existing packages

### DISCUSSION THREAD

**Discussion by sarbagyastha** [sarbagyastha/youtube_player_flutter#647] [Ideas]
The refactor will include following changes:

- [x] Work on improving the code structure.
- [x] Resolve reported issues.
- [x] Make `youtube_player_flutter` dependent on `youtube_player_iframe`'s API.
- [x] Switch to official [`webview_flutter`](https://pub.dev/packages/webview_flutter), now that it seems to include all the APIs required, which was not there before, alongside the web support.
- [ ] Add test cases

**Comment by mark8044** [comment]
YES!
Glad to see this repo coming back to life.  Im excited to see the updates happening.

T H A N K   Y O U

**Comment by mark8044** [comment]
Also is it possible to do anything about iframe accepted clicks even when the video is covered by another view?

**Reply by sarbagyastha** [reply]
For mobile, gesture recognizers can be used. 
For web, I don't think it is possible yet. But will look into it.

**Reply by sarbagyastha** [reply]
For web, this should work, https://pub.dev/packages/pointer_interceptor

**Comment by sarbagyastha** [comment]
Released  [youtube_player_iframe **3.0.0-beta.1**](https://pub.dev/packages/youtube_player_iframe/versions/3.0.0-beta.1/changelog)

**Reply by mark8044** [reply]
> Released [youtube_player_iframe **3.0.0-beta.1**](https://pub.dev/packages/youtube_player_iframe/versions/3.0.0-beta.1/changelog)

Changes look great so far and things *seem* a little smoother?

With `initialVideoId` now gone, is the proper way to load an initial video, to cue it via the controller?  Im doing it that way for now and its working, but wanted to make sure that's the suggested way

```
OLD:

  @override
  Widget build(BuildContext context) {

    YoutubePlayerController controller = YoutubePlayerController(
      initialVideoId: videoId,
      params: const YoutubePlayerParams(
          showControls: true,
          showFullscreenButton: true,
      ),
    );


    return YoutubePlayerIFrame(
      controller: controller,
      aspectRatio: 16 / 9,
    );
    
  }
```

```
NEW:

  @override
  Widget build(BuildContext context) {

    YoutubePlayerController controller = YoutubePlayerController(
      params: const YoutubePlayerParams(
          showControls: true,
          showFullscreenButton: true,
      ),
    );

    controller.cueVideoById(videoId: videoId);

    return YoutubePlayerIFrame(
      controller: controller,
      aspectRatio: 16 / 9,
    );
    
  }

```

**Reply by sarbagyastha** [reply]
> With initialVideoId now gone, is the proper way to load an initial video, to cue it via the controller? 

Yes, you can use `cue` or `load` methods.

But, you should not use in inside `build` method directly, as it'll cue the same video multiple times.

**Reply by habibrosyad** [reply]
I try the example for v3, but it is not working, no video shows up at all.

**Reply by habibrosyad** [reply]
Also, as far as I know, similar package like https://github.com/yh-luo/webviewtube which use webview_flutter has bugs, which cannot work with youtube video with id containing underscore (e.g. 'i_ZrWS8wMMI'). I tried to call the IFrame API inside the webview, it seems to be the problem with webview_flutter, it works just fine in flutter_inappwebview.

**Reply by sarbagyastha** [reply]
Published `youtube_player_iframe: 3.0.0-beta.2` with the fix.

**Reply by CosmicPangolin** [reply]
@sarbagyastha 

Thanks for working on the package.

It strikes me as strange that we can't assign an initial value in the controller declaration...most Flutter controllers provide this in the API...and the onInit assignment pattern you show is unlike anything I've seen used in this language.  The functionality feels hidden since the variable isn't exposed as a constructor field?  This new API makes my code more verbose for the same functionality :/ - 

![image](https://user-images.githubusercontent.com/40650466/184789249-7588d76d-948b-4348-84bc-58e4eeea8d78.png)

**Reply by sarbagyastha** [reply]
Added `YoutubePlayerController.fromVideoId` with v3.0.4

**Comment by deepak786** [comment]
```
strictRelatedVideos: true,
```

seems not to be working on v3.
